File: src/lib/annis/operators/overlap.cpp
/*
   Copyright 2017 Thomas Krause <thomaskrause@posteo.de>

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

#include "overlap.h"
#include <annis/wrapper.h>                    // for ListWrapper
#include <google/btree_set.h>                 // for btree_set
#include <utility>                            // for pair, move
#include <vector>                             // for vector
#include "annis/db.h"                         // for DB
#include "annis/graphstorage/graphstorage.h"  // for ReadableGraphStorage
#include "annis/graphstorageholder.h"         // for GraphStorageHolder
#include "annis/iterators.h"                  // for EdgeIterator, AnnoIt
#include "annis/operators/operator.h"         // for Operator
#include "annis/util/helper.h"                // for TokenHelper


using namespace annis;

Overlap::Overlap(const DB &db, GraphStorageHolder& gsh)
  : tokHelper(gsh, db), anyNodeAnno(Init::initAnnotation(db.getNodeNameStringID(), 0, db.getNamespaceStringID()))
{
  gsOrder = gsh.getGraphStorage(ComponentType::ORDERING, annis_ns, "");
  gsCoverage = gsh.getGraphStorage(ComponentType::COVERAGE, annis_ns, "");
  gsInverseCoverage = gsh.getGraphStorage(ComponentType::INVERSE_COVERAGE, annis_ns, "");

}

std::unique_ptr<AnnoIt> Overlap::retrieveMatches(const annis::Match &lhs)
{
  std::unique_ptr<ListWrapper> w = std::unique_ptr<ListWrapper>(new ListWrapper());

  btree::btree_set<nodeid_t> uniqueResultSet;

  // get covered token of lhs
  if(tokHelper.isToken(lhs.node))
  {
    // get all nodes that are covering this token
    std::vector<nodeid_t> overlapCandidates = gsInverseCoverage->getOutgoingEdges(lhs.node);
    for(const auto& c : overlapCandidates)
    {
      uniqueResultSet.insert(c);
    }
     // also add the token itself
    uniqueResultSet.insert(lhs.node);
  }
  else
  {

    std::unique_ptr<EdgeIterator> coveredByLeftIt
        = gsCoverage->findConnected(lhs.node);
    for(auto leftToken = coveredByLeftIt->next();
        leftToken.first; leftToken = coveredByLeftIt->next())
    {

      // get all nodes that are covering the token
      std::vector<nodeid_t> overlapCandidates = gsInverseCoverage->getOutgoingEdges(leftToken.second);
      for(const auto& c : overlapCandidates)
      {
        uniqueResultSet.insert(c);
      }
       // also add the token itself
      uniqueResultSet.insert(leftToken.second);
    }
  }

  // add all unique matches to result
  for(const auto& m : uniqueResultSet)
  {
    w->addMatch(Init::initMatch(anyNodeAnno, m));
  }

  return std::move(w);
}

bool Overlap::filter(const Match &lhs, const Match &rhs)
{
  nodeid_t lhsLeftToken = tokHelper.leftTokenForNode(lhs.node);
  nodeid_t lhsRightToken = tokHelper.rightTokenForNode(lhs.node);
  nodeid_t rhsLeftToken = tokHelper.leftTokenForNode(rhs.node);
  nodeid_t rhsRightToken = tokHelper.rightTokenForNode(rhs.node);

  if(gsOrder->distance(Init::initEdge(lhsLeftToken, rhsRightToken)) >= 0
     && gsOrder->distance(Init::initEdge(rhsLeftToken, lhsRightToken)) >= 0)
  {
    return true;
  }
  return false;
}

double Overlap::selectivity() 
{
  if(gsOrder == nullptr || gsCoverage == nullptr)
  {
    return Operator::selectivity();
  }

  auto statsCov = gsCoverage->getStatistics();
  auto statsOrder = gsOrder->getStatistics();


  double numOfToken = statsOrder.nodes;

  if(statsCov.nodes == 0)
  {
    // only token in this corpus
    return 1.0 / numOfToken;
  }
  else
  {

    // Assume two nodes have overlapping coverage if the left- or right-most covered token is inside the
    // covered range of the other node.
    return ((statsCov.avgFanOut*2.0) / numOfToken);
  }

}


Overlap::~Overlap()
{

}

File: graphannis/src/operator/overlap.rs
use super::{Operator, OperatorSpec};
use {Annotation, Component, ComponentType, Match, NodeID};
use graphdb::GraphDB;
use graphstorage::GraphStorage;
use util::token_helper;
use util::token_helper::TokenHelper;
use std::collections::HashSet;

use std::rc::Rc;
use std;

#[derive(Clone)]
pub struct OverlapSpec;

pub struct  Overlap<'a> {
    gs_order: Rc<GraphStorage>,
    gs_cov: Rc<GraphStorage>,
    gs_invcov: Rc<GraphStorage>,
    tok_helper: TokenHelper<'a>,
}

lazy_static! {

    static ref COMPONENT_ORDER : Component =  {
        Component {
            ctype: ComponentType::Ordering,
            layer: String::from("annis"),
            name: String::from(""),
        }
    };


    static ref COMPONENT_COVERAGE : Component =  {
        Component {
            ctype: ComponentType::Coverage,
            layer: String::from("annis"),
            name: String::from(""),
        }
    };

    static ref COMPONENT_INV_COVERAGE : Component =  {
        Component {
            ctype: ComponentType::InverseCoverage,
            layer: String::from("annis"),
            name: String::from(""),
        }
    };
}

impl OperatorSpec for  OverlapSpec {
    fn necessary_components(&self) -> Vec<Component> {
        let mut v: Vec<Component> = vec![COMPONENT_ORDER.clone(), 
            COMPONENT_COVERAGE.clone(), COMPONENT_INV_COVERAGE.clone()];
        v.append(&mut token_helper::necessary_components());
        v
    }

    fn create_operator<'b>(&self, db: &'b GraphDB) -> Option<Box<Operator + 'b>> {
        let optional_op =  Overlap::new(db);
        if let Some(op) = optional_op {
            return Some(Box::new(op));
        } else {
            return None;
        }
    }
}

impl<'a>  Overlap<'a> {
    pub fn new(db: &'a GraphDB) -> Option< Overlap<'a>> {
        let gs_order = db.get_graphstorage(&COMPONENT_ORDER)?;
        let gs_cov = db.get_graphstorage(&COMPONENT_COVERAGE)?;
        let gs_invcov = db.get_graphstorage(&COMPONENT_INV_COVERAGE)?;        

        let tok_helper = TokenHelper::new(db)?;

        Some( Overlap {
            gs_order,
            gs_cov,
            gs_invcov,
            tok_helper,
        })
    }
}

impl<'a> std::fmt::Display for Overlap<'a> {
     fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "_o_")
    }
}

impl<'a> Operator for  Overlap<'a> {
    fn retrieve_matches<'b>(&'b self, lhs: &Match) -> Box<Iterator<Item = Match> + 'b> {

        // use set to filter out duplicates
        let mut result = HashSet::new();

        let covered : Box<Iterator<Item=NodeID>> = if self.tok_helper.is_token(&lhs.node) {
            Box::new(std::iter::once(lhs.node))
        }  else {
            // all covered token
            Box::new(self.gs_cov.find_connected(&lhs.node, 1, 1))
        };

        for t in covered {
             // get all nodes that are covering the token
            for n in self.gs_invcov.find_connected(&t, 1, 1) {
                result.insert(n);
            }
            // also add the token itself
            result.insert(t);
        }

        return Box::new(result.into_iter().map(|n| Match{node: n, anno : Annotation::default()}));
    }

    fn filter_match(&self, lhs: &Match, rhs: &Match) -> bool {
        if let (Some(start_lhs), Some(end_lhs), Some(start_rhs), Some(end_rhs)) = (
            self.tok_helper.left_token_for(&lhs.node),
            self.tok_helper.right_token_for(&lhs.node),
            self.tok_helper.left_token_for(&rhs.node),
            self.tok_helper.right_token_for(&rhs.node),
        ) {
            // TODO: why not isConnected()? (instead of distance)
            // path between LHS left-most token and RHS right-most token exists in ORDERING component
            if self.gs_order.distance(&start_lhs, &end_rhs).is_some()
                // path between LHS left-most token and RHS right-most token exists in ORDERING component
                && self.gs_order.distance(&start_rhs, &end_lhs).is_some() {
                    return true;
            }
        }
        return false;
    }
}
